### PR TITLE
*: make changes to tidb_schema_cache_size effective immediately

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref #50959

Problem Summary:

Previously, if we change `tidb_schema_cache_size` to a different number, we still need to execute something(like DDL) to trigger `loadInfoSchema` to make it effective.

### What changed and how does it work?

This PR makes changes to tidb_schema_cache_size effective immediately.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  image
  ```
  mysql> set global tidb_schema_cache_size = 1000;
  Query OK, 0 rows affected (0.02 sec)
  
  mysql> set global tidb_schema_cache_size = 10000;
  Query OK, 0 rows affected (0.04 sec)
  
  mysql> set global tidb_schema_cache_size = 100000;
  Query OK, 0 rows affected (0.04 sec)
  ```
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/infoschema/sieve.go [hunk 1]
```diff
@@ -75,6 +75,14 @@ func (s *Sieve[K, V]) SetCapacity(capacity uint64) {
 	s.capacity = capacity
 }
 
+func (s *Sieve[K, V]) SetCapacityAndWaitEvict(capacity uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for s.size > s.capacity {
+		s.evict()
```
Comment (tangenta) [inline pkg/infoschema/sieve.go:82]: This may block the access to cache if the difference is large, is there any suggestion for this risk?

Comment (lance6716) [reply pkg/infoschema/sieve.go:82]: maybe use a loop and in each round, acuqire the lock, evict some number of items (like in line 102 ~ 104), release the lock.

**Review by wjhuang2016 — APPROVED** [review]

File: pkg/infoschema/sieve.go [hunk 2]
```diff
@@ -75,6 +75,17 @@ func (s *Sieve[K, V]) SetCapacity(capacity uint64) {
 	s.capacity = capacity
 }
 
+func (s *Sieve[K, V]) SetCapacityAndWaitEvict(capacity uint64) {
```
Comment (lance6716) [inline pkg/infoschema/sieve.go:78]: I'm not sure if there will be two call, one wants to increase capacity and the other wants to decrease. Maybe we should add another lock to make sure `SetCapacityAndWaitEvict` is not concurrently executing.

**Comment by lance6716** [comment]
rest lgtm

File: pkg/infoschema/sieve.go (hunk 2 above)
Comment (tangenta) [reply pkg/infoschema/sieve.go:78]: Do you mean the access of `s.size` and `s.capacity` is not protected by lock?

**Review by lance6716 — APPROVED** [review]

**Comment by tiancaiamao** [comment]
This works when it's infoschema v2 already ...
But in case it's infoschema v1 currently, that change does not make it run v2 immediately.

**Pull request merged**